core/views.py

```diff
@@ -0,0 +1,256 @@
+                raw_password = form.cleaned_data.get("password1")
+                user = authenticate(username=username, password=raw_password)
+                auth_login(request, user)
+                hacker = hacker_models.Hacker.objects.get(username=username)
+                # call the 'send_confirmation_email' function to send a confirmation email
+                self.send_confirmation_email(hacker)
+                return redirect(reverse_lazy("confirm_email"))
+            else:
+                user_exists = True
+        FormErrors = json.loads(form.errors.as_json())
+        return render(
+            request,
+            self.template_name,
+            {"form": form, "FormErrors": FormErrors, "user_exists": user_exists},
+        )
+
+
+# django.contrib.auth.LoginView
+class SignInView(generic_views.FormView):
+    form_class = hacker_forms.SignInForm
+    template_name = "registration/login.html"
+    success_url = hacker_forms.SignInForm.success_url  #!!!
+    redirect_field_name = hacker_forms.SignInForm.redirect_field_name
+
+    # references: https://gist.github.com/stefanfoulis/1140136 , https://coderwall.com/p/sll1kw/django-auth-class-based-views-login-and-logout
+    @method_decorator(csrf_protect)
+    @method_decorator(never_cache)
+    def dispatch(self, *args, **kwargs):
+        return super(SignInView, self).dispatch(*args, **kwargs)
+
+    def form_valid(self, form):
+        """
+        The user has provided valid credentials (this was checked in AuthenticationForm.is_valid()). So now we
+        can log them in.
+        """
+        auth_login(self.request, form.get_user())
+        return redirect(self.get_success_url())
+
+    def form_invalid(self, form):
+        return render(self.request, self.template_name, {"invalid_form": True})
+
+    def get_success_url(self):  #!!!!!!!!!
+        return self.success_url
+
+    def set_test_cookie(self):
+        self.request.session.set_test_cookie()
+
+    def check_and_delete_test_cookie(self):
+        if self.request.session.test_cookie_worked():
+            self.request.session.delete_test_cookie()
+            return True
+        return False
+
+    def get(self, request, *args, **kwargs):  #!!!!!!!!!
+        if request.user.is_authenticated:
+            return redirect(self.get_success_url())
+        form = self.form_class(initial=self.initial)
+        self.set_test_cookie()
+        return super(SignInView, self).get(request, *args, **kwargs)
+
+    @method_decorator(csrf_protect)
+    def post(self, request, *args, **kwargs):
+        form = self.form_class(request.POST)
+
+        form_class = self.get_form_class()
+        form = self.get_form(form_class)
+
+        if form.is_valid():
+            self.check_and_delete_test_cookie()
+            return self.form_valid(form)
+        else:
+            self.set_test_cookie()
+            return self.form_invalid(form)
+
+
+class LogOutView(RedirectView):
```

Comment:
See the comment on `SignInView`. We don't need this class, we can just use Django's built-in [`LogoutView`](https://docs.djangoproject.com/en/dev/topics/auth/default/#django.contrib.auth.views.LogoutView) in the URLs.